File: Semestrul 4/MPP/Proiecte/java_networking/Networking/src/main/java/networking/rpcprotocol/ServicesRpcProxy.java
package networking.rpcprotocol;

import model.*;
import services.IProjectObserver;
import services.IProjectServices;

import java.io.EOFException;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

public class ServicesRpcProxy implements IProjectServices {
    private String host;
    private int port;

    private IProjectObserver client;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;

    private BlockingQueue<Response> qresponses;
    private volatile boolean finished;
    public ServicesRpcProxy(String host, int port){
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<Response>();
    }

    private void initializeConnection() throws Exception {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (IOException e){
            throw new Exception("Error connecting to the server: " + e.getMessage());
        }
    }

    private void closeConnection(){
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) throws Exception{
        if (output == null) initializeConnection();

        try {
            output.writeObject(request);
            output.flush();
        }catch (IOException e){
            throw new Exception("Error sending object: " + e.getMessage());
        }
    }

    private Response readResponse() throws Exception{
        try {
            return qresponses.poll(30, TimeUnit.SECONDS);
        }catch (InterruptedException e){
            throw new Exception("Error reading response: " + e.getMessage());
        }
    }

    private void startReader() throws Exception{
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private boolean isUpdate(Response response) {
        // Tratază ca update doar dacă tipul este DONATIE_NOUA

        return response.type() == ResponseType.DONATIE_NOUA;
    }

    private void handleUpdate(Response response){
        if (isUpdate(response)){
            Donatie dono = (Donatie) response.data();
            try {
                client.adauga(dono);
            }catch (Exception e){
                e.printStackTrace();
            }
        }
    }

    private class ReaderThread implements Runnable{
        public void run(){
            while (!finished){
                try {
                    Object response = input.readObject();
                    if (isUpdate((Response) response)){
                        handleUpdate((Response) response);
                    }
                    else {
                        qresponses.put((Response) response);
                    }
                }catch (SocketException | EOFException e){
                    System.err.println("Socket closed, stopping reader");
                    break;
                }catch (IOException | ClassNotFoundException | InterruptedException e){
                    e.printStackTrace();
                }
            }
        }
    }

    @Override
    public Voluntar login(String username, String password, IProjectObserver client) throws Exception {
        initializeConnection();
        this.client = client;
        Voluntar voluntar = new Voluntar(username, password);
        Request req = new Request.Builder().type(RequestType.LOGIN).data(voluntar).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.OK){
            return (Voluntar) response.data();
        }
        throw new Exception(response.data().toString());
    }

    // În ServicesRpcProxy.java
    @Override
    public Donatie saveDonatie(Donatie donatie) throws Exception {
        System.out.println("Trimitere cerere DONATIE_NOUA: " + donatie);
        Request req = new Request.Builder().type(RequestType.DONATIE_SAVE).data(donatie).build();
        sendRequest(req);
        Response response = readResponse();

        if (response == null) {
            throw new Exception("Nu s-a primit răspuns de la server.");
        }

        System.out.println("Răspuns primit: " + response);
        if (response.type() == ResponseType.ERROR) {
            throw new Exception(response.data().toString());
        }
        return (Donatie) response.data();
    }

    @Override
    public int getSumaDonatiiPentruCaz(int cazId) throws Exception {
        Request req = new Request.Builder().type(RequestType.GET_SUMA_DONATII_PT_CAZ).data(cazId).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR){
            throw new Exception(response.data().toString());
        }

        return (Integer) response.data();
    }

    @Override
    public List<Caz> findAllCaz() throws Exception {
        Request req = new Request.Builder().type(RequestType.FIND_ALL_CAZ).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR)
            throw new Exception(response.data().toString());

        return (List<Caz>) response.data();
    }

    // În ServicesRpcProxy
    @Override
    public List<CazDTO> createCazDTOList() throws Exception {
        Request req = new Request.Builder().type(RequestType.LIST_DTO_CAZ).build();
        sendRequest(req);
        Response response = readResponse();
        if (response == null) {
            throw new Exception("Nu s-a primit răspuns de la server.");
        }
        if (response.type() != ResponseType.LIST_DTO_CAZ) { // Verifică tipul răspunsului
            throw new Exception("Răspuns neașteptat de la server: " + response.type());
        }
        return (List<CazDTO>) response.data();
    }

    @Override
    public List<Donator> searchByName(String name) throws Exception {
        Request req = new Request.Builder().type(RequestType.SEARCH_DONO_BYPNAME).data(name).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR)
            throw new Exception(response.data().toString());

        return (List<Donator>) response.data();
    }

    @Override
    public Donator saveDonator(Donator donator) throws Exception {
        Request req = new Request.Builder().type(RequestType.SAVE_DONO).data(donator).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR)
            throw new Exception(response.data().toString());

        return (Donator) response.data();
    }

    @Override
    public Donator findByFullName(String fullName) throws Exception {
        Request req = new Request.Builder().type(RequestType.FIND_DONO).data(fullName).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR)
            throw new Exception(response.data().toString());

        return (Donator) response.data();
    }

    @Override
    public void logout(String username) throws Exception {
        Request req = new Request.Builder()
                .type(RequestType.LOGOUT)
                .data(username)
                .build();
        sendRequest(req);
        Response response = readResponse(); // dacă nu faci pe thread separat, aici se poate bloca!

        if (response.type() == ResponseType.ERROR) {
            throw new Exception("Logout failed: " + response.data());
        }
    }

}

File: Semestrul 4/MPP/Proiecte/c#_networking/networkingnet/RpcProtocol/ServicesRpcProxy.cs
using System.Text.Json.Serialization;
using servicesnet;

namespace networkingnet.RpcProtocol;

using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using modelnet;
using Microsoft.VisualBasic.CompilerServices;

public class ServicesRpcProxy : IProjectServices
{
    private string host;
    private int port;

    private IProjectObserver client;
    private NetworkStream stream;
    private TcpClient connection;
    private Queue<Response> responses;
    private volatile bool finished;
    private EventWaitHandle waitHandle;

    // ServicesRpcProxy.cs
    private static readonly JsonSerializerOptions jsonOptions = new JsonSerializerOptions
    {
        PropertyNameCaseInsensitive = true, // Ignoră diferențele de casing
        IncludeFields = true, // Include câmpuri dacă sunt folosite
        WriteIndented = true,
        // Adaugă suport pentru constructori parametrizați
        DefaultIgnoreCondition = JsonIgnoreCondition.Never
    };
    public ServicesRpcProxy(string host, int port)
    {
        this.host = host;
        this.port = port;
        responses = new Queue<Response>();
        try
        {
            initializeConnection(); // Inițializează conexiunea la server
        }
        catch (ProjectException e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    private void sendRequest(Request request)
    {
        try
        {
            lock (stream)
            {
                string json = JsonSerializer.Serialize(request);
                //log.DebugFormat("Sending json request {0}", json);
                byte[] bytes = Encoding.UTF8.GetBytes(json + "\n");
                stream.Write(bytes, 0, bytes.Length);
                stream.Flush();
            }
        }
        catch (Exception e)
        {
            //log.ErrorFormat("Error sending request {0}", e.Message);
            if (e.InnerException != null)
                // log.ErrorFormat("Error sending request inner error {0}", e.InnerException.Message);
                throw new ProjectException("Error sending request " + e);
        }
    }

    private Response readResponse()
    {
        Response response = null;
        try
        {
            waitHandle.WaitOne();
            lock (responses)
            {
                response = responses.Dequeue();
            }
        }
        catch (Exception e)
        {
            if (e.InnerException != null) ;
        }

        return response;
    }

    private void initializeConnection()
    {
        try
        {
            connection = new TcpClient(host, port);
            stream = connection.GetStream();
            finished = false;
            waitHandle = new AutoResetEvent(false);
            startReader();
        }
        catch (Exception e)
        {
            //log.ErrorFormat("Error initializing connection {0}", e.Message);
            if (e.InnerException != null) ;
            //log.ErrorFormat("Error initializing connection inner error {0}", e.InnerException.Message);
            Console.WriteLine("Error initializing connection " + e.Message);
        }
    }

    private void closeConnection()
    {
        finished = true;
        try
        {
            stream.Close();
            connection.Close();
            waitHandle.Close();
            client = null;
        }
        catch (Exception e)
        {
            //log.ErrorFormat("Error closing connection {0}", e.Message);
            if (e.InnerException != null) ;
            //log.ErrorFormat("Error closing connection inner error {0}", e.InnerException.Message);
        }
    }

    private void startReader()
    {
        Thread thread = new Thread(run);
        thread.Start();
    }

    private void handleUpdate(Response response)
    {
        //log.DebugFormat("Received json response {0}", JsonSerializer.Serialize(response));
        if (response.Type == ResponseType.DONATIE_NOUA)
        {
            try
            {
                //log.Debug($"Attempting to notify client with ticket: {response.Bilet != null}");
                if (client == null)
                {
                    Console.WriteLine("Client is null when trying to notify about ticket");
                    return;
                }

                // Use a try-catch here as the client might be disconnected
                try
                {
                    client.notifyObservers(response.Donatie);
                }
                catch (Exception ex)
                {
                    Console.WriteLine("Failed to notify client: {ex.Message}");
                }
            }
            catch (Exception e)
            {
                // log.ErrorFormat("Error handling update {0}", e.Message);
                if (e.InnerException != null) ;
                // log.ErrorFormat("Error handling update inner error {0}", e.InnerException.Message);
            }
        }
    }

    private bool isUpdate(Response response)
    {
        return response.Type == ResponseType.DONATIE_NOUA;
    }

    private void run()
    {
        using StreamReader streamReader = new StreamReader(stream, Encoding.UTF8);
        while (!finished)
        {
            try
            {
                string json = streamReader.ReadLine();
                if (string.IsNullOrEmpty(json))
                    continue;
                //log.DebugFormat("Received json response {0}", json);
                Response response = JsonSerializer.Deserialize<Response>(json);
                //log.DebugFormat("Deserialized response {0}", JsonSerializer.Serialize(response));
                if (isUpdate(response))
                {
                    handleUpdate(response);
                }
                else
                {
                    lock (responses)
                    {
                        responses.Enqueue(response);
                    }

                    waitHandle.Set();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Error reading response {0}", e.Message);
                if (e.InnerException != null) ;
                //log.ErrorFormat("Error reading response inner error {0}", e.InnerException.Message);
            }
        }
    }

    public Voluntar login(string username, string password, IProjectObserver client)
    {
        Console.WriteLine("Login...");
        sendRequest(JsonProtocolUtils.createLoginRequest(username, password));
        Response response = readResponse();
        if (response.Type == ResponseType.OK && response.Voluntar != null)
        {
            this.client = client;
            Console.WriteLine("Login successful");
            return response.Voluntar;
        }
        string err = response.ErrorMessage ?? "Unknown login error";
        closeConnection();
        throw new ProjectException(err);
        
    }
    

    public Donatie saveDonatie(Donatie donatie)
    {
        sendRequest(JsonProtocolUtils.createSaveDonatie(donatie));
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }

        return response.Donatie;

    }

    public List<Caz> findAllCaz()
    {
        sendRequest(JsonProtocolUtils.createFindAllCaz());
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }

        List<Caz> cazuri = response.Cazuri.ToList();
        return cazuri;
    }

    public List<CazDTO> createCazDTOList()
    {
        sendRequest(JsonProtocolUtils.createListDTOCaz());
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }
        List<CazDTO> cazuriDTO = response.CazuriDTO.ToList();
        return cazuriDTO;
    }

    public List<Donator> searchByName(string partialName)
    {
        sendRequest(JsonProtocolUtils.createSearchDonoByPName(partialName));
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }
        List<Donator> donatori = response.Donatori.ToList();
        return donatori;
    }
    

    public Donator saveDonator(Donator donator)
    {
        sendRequest(JsonProtocolUtils.createSaveDonator(donator));
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }
        return response.Donator;
    }

    public Donator findByFullName(string numeDonator)
    {
        sendRequest(JsonProtocolUtils.createFindDonator(numeDonator));
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }
        return response.Donator;
    }

    public int getSumaDonatiiPentruCaz(int cazId)
    {
        sendRequest(JsonProtocolUtils.createGetSumaDonatiiPtCaz(cazId));
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }

        return response.SumaPentruCaz;
    }

    public void logout(string username, IProjectObserver client)
    {
        sendRequest(JsonProtocolUtils.CreateLogoutRequest(username));
        Response response = readResponse();
        closeConnection();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }
    }

    public IEnumerable<Voluntar> findAllVoluntari()
    {
        sendRequest(JsonProtocolUtils.createfindAllVoluntariRequest());
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR)
        {
            throw new ProjectException("Error sending request");
        }

        return response.Voluntari;
    }
}